Connection api fixes

vspinu opened the pull request:
A range of simple comits which should fix #2350, fix #2349, fix #2342 and fix #2340.

I am strugling with cask install locally. For unclear reason it pulls an one week old sesman. Both MELPA and MELPA-stable have up to date sesman. Any ideas?

vspinu commented:
Same happens on travis. Old sesman is pulled there as well.

vspinu commented:
Looks like it's a melpa issue actually (melpa/melpa#5580).  I was mislead by the github badge which was up to date.

bbatsov commented on cider-repl.el:
```diff
@@ -1027,10 +1027,12 @@ text property `cider-old-input'."
+         (other-repl (car (seq-remove (lambda (r) (eq r this-repl)) (cider-repls nil t)))))
+    (if other-repl
+        (switch-to-buffer other-repl)
+      (user-error "No other REPL in current session (%s)"
```
I think it'd be much better to use the type of the other REPL in this message, instead of the vague word "other". E.g. "No ClojureScript REPL...".

vspinu replied:
This function is a bit more generic now. It will work with sessions containing two cljs connections for example. So the other here literally means "other repl besides the current one". 

This function should be fixed to cycle through all of the repls in case there are more than two. Will fix that in due time.

bbatsov commented on cider-connection.el:
```diff
@@ -517,32 +516,42 @@ function with the repl buffer set as current."
 ;;; Current/other REPLs
 
-(defun cider-current-repl (&optional type)
+(defun cider-current-repl (&optional type ensure)
```
I'm puzzled by this change. Can you even have a REPL without a session?

bbatsov replied:
P.S. I'd also name this param `ensure-session`.

vspinu replied:
Ensure means that it will throw an error if there is no repl. And it means that it will throw both when there is no session and when there is no repl of type TYPE within a session. We cannot have sessions with no repls, but we can have sessions consisting of, for example, clj repls only.

bbatsov replied:
Btw, why is the type a string instead of a symbol? That seems a bit unlispy to me.

vspinu replied:
:) that's because `cider-repl-type` returns a string, which in turn (my guess) is because the state handler returns a type as a string. So I just don't want to convert things while passing around. 

It might be good idea to change the type to a keyword throughout. It's a keyword in `cider-map-connections` argument.

bbatsov commented on cider-connection.el:
```diff
@@ -337,7 +336,7 @@ Don't restart the server or other connections within the same session.  Use
   "Display information about the current connection."
   (interactive)
-  (message "%s" (cider--connection-info (cider-current-repl))))
+  (message "%s" (cider--connection-info (cider-current-repl nil t))))
```
I prefer passing symbols like `'ensure-session` instead of `t` for such boolean params. That makes the code much more readable.

vspinu replied:
Sure, will change. But I am a bit divided on this one.  For frequently used functions like `cider-current-repl` everyone will know what 2nd parameter means, so adding `'ensure` just eats space for nothing.

bbatsov commented on cider.el:
```diff
@@ -1177,7 +1177,7 @@ choose."
 ;; TODO: Implement a check for command presence over tramp
 (defun cider--resolve-command (command)
-  "Find COMMAND on `exec-path' if possible, or return nil.
+  "Find COMMAND in list variable ‘exec-path’ if available, or return nil.
```
You'll have to explain to me this change. Seems pretty weird to me.

vspinu replied:
It is, but checkdoc insists to disambiguate `exec-path` with `variable` or `function` in front. Pretty silly but I don't see how this could be done otherwise without disabling the check or removing `exec-path`. "list variable `exec-path` is a bit more meaningful than `in variable exec-path`.  An alternative would be to just write `in exec path`.

bbatsov commented:
So, we'll have to wait until the proper build is generated, right?

vspinu commented on cider-connection.el:
```diff
@@ -587,6 +581,64 @@ Error is signaled if no REPL buffer of specified type exists."
+
+;; REPLs double as connections in CIDER, so it's useful to be able to refer to
+;; them as connections in certain contexts.
+(defalias 'cider-current-connection #'cider-current-repl)
```
These 4 for a unit of related functionality with the inline comment above relating to all of them. Plus all of these functions are defined right there an inch above.

vspinu commented on cider-connection.el:
```diff
@@ -587,6 +581,64 @@ Error is signaled if no REPL buffer of specified type exists."
+(define-obsolete-function-alias 'nrepl-connection-buffer-name 'nrepl-repl-buffer-name "0.18")
+(define-obsolete-function-alias 'cider-repl-set-type 'cider-set-repl-type "0.18")
+
+(make-obsolete 'cider-assoc-buffer-with-connection 'sesman-link-with-buffer "0.18")
```
I am not sure what you mean here, but `make-obsolete` is used here for all functions which don't have one-to-one mapping into new functionality.

vspinu commented:
Yes. Once MELPA is on I will re-base with your suggestions.

vspinu commented:
MELPA is back. I have addressed your comments except for the location of connection aliases. It's the most natural grouping in my mind, but feel free to move around as you see fit. They won't be aliases much longer though, if follow with the connection <=> process refactoring.

Tests fail due to a bug in clojure-mode. Should be fine once fixed.

bbatsov commented:
> They won't be aliases much longer though, if follow with the connection <=> process refactoring.

Yeah, that'd be best. Btw, now that sesman is standalone, please don't forget to make me a co-owner of the project.
